DSDEGP-1812 Generalize existing move logic to allow setting arbitrary new basename

danxmoran opened the pull request:
### Description

This is needed for reset-ing filenames after testing a cram delivery with a patient-ID prefix, and is also useful in general.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad commented on clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/MoveGvcfSpec.scala:
```diff
@@ -19,7 +19,8 @@ class MoveGvcfSpec extends BaseClientSpec {
         key = testGvcfTransferV1Key,
-        destination = MockIoUtil.InvalidPath
+        destination = MockIoUtil.InvalidPath,
+        None
```
could you use an overloaded constructor to avoid having to pass `None`?

danxmoran replied:
I can probably use a default value. Overloads don't play nicely with case classes AFAIK.

pshapiro4broad commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverWgsCramExecutor.scala:
```diff
@@ -27,16 +27,24 @@ import scala.concurrent.{ExecutionContext, Future}
-    val moveExecutor =
-      new MoveExecutor(moveCommand, deliverCommand.samplePrefix)
+    val deliveredFileBasename: String =
+      deliverCommand.samplePrefix.fold(deliverCommand.key.sampleAlias) {
```
Did you consider accepting a function that could transform the destination filename, instead of the new basename? I think it would make the code here slightly simpler.

danxmoran replied:
Not sure I understand... where would the function live? Right now `deliverCommand.samplePrefix` is coming from the command line, so I think it's stuck as an `Option[String]`. We need to perform this computation somewhere, so I'm not sure wrapping it in a function to be called by the methods in the `TransferMetadata` would be any simpler than calling it once here and passing in the result.

pshapiro4broad replied:
The function is a lambda created at the call site. Does the caller have enough information to replace the basename without needing the extension? If that could be avoided it would be much better. I don't know if we can be sure extensions of these files will never change in the future, and hardcoding them here doesn't seem great to me.

danxmoran replied:
Ah! I think I get what you're suggesting now, but I still don't know how it'd work.

In the special case of cram delivery, the caller knows that it's manipulating files that should end in ".cram", ".crai", etc, but in general that's not the case. The move-with-new-basename functionality is built into the move executor, which knows very little about the files it's operating over, and the new basename is getting parsed straight from the CLI. There's no special-cased intermediate between the parser and the executor that could transform the new basename into a lambda, and no good way to specify the lambda directly on the command line.

I managed to convince myself yesterday that the least-bad way to implement this feature is to hardcode the expected extensions. The concrete situation Sam was trying to recover from was a bug in the delivery WDL that ended up naming things like `..NA12878.cram`. I couldn't think of a sane splitting algorithm that would handle cases like this along with:

- `NA12878.cram`
- `SomePatientId.NA12878.cram`
- `NA12878.cram.crai`
- `HK3T5CCXX.6.Pond-492100.unmapped.bam`
- `NA12878.v10.bb684530-32a6-4461-a9f5-84778f61d935.g.vcf.gz`

Since I had convinced myself that the only way to tell basename apart from extension would be to hardcode the extensions somewhere, I stuck them in the transfer-metadata classes, which is where we've put all the other index-specific logic.

danxmoran replied:
But I agree there are definitely drawbacks to having things hard-coded. I built a jar from my branch and gave it to Sam to unblock him, so we can leave this open over the break and talk about it with the whole group to see if this is something we want to maintain long-term.

tbl3rd replied:
Type these members not as URIs but some ExendedUri type with an extension?

danxmoran replied:
Made [DSDEGP-1827](https://broadinstitute.atlassian.net/browse/DSDEGP-1827) to track it.

pshapiro4broad commented on clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/GvcfTests.scala:
```diff
@@ -570,10 +572,10 @@ trait GvcfTests { self: BaseIntegrationSpec =>
+    val gvcfName = s"$randomId.g.vcf.gz"
+    val indexName = s"$randomId.g.vcf.gz.tbi"
+    val summaryMetricsName = s"$randomId.variant_calling_summary_metrics"
+    val detailMetricsName = s"$randomId.variant_calling_detail_metrics"
```
Even though this is test code, I think we want apply the same rigor to it regarding code duplication. Can the common test setup code be refactored into a common method?

danxmoran replied:
Done! Glorious -642 lines

pshapiro4broad replied:
![bitmoji](https://render.bitstrips.com/v2/cpanel/3c97455d-0ce1-4ac8-82a0-601e6d6f6fd2-58bd1e74-5b74-4b76-b7ec-5ecb58ede85e-v1.png?transparent=1&palette=1&width=246)

pshapiro4broad commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/TransferWgsCramV1Metadata.scala:
```diff
@@ -70,19 +69,16 @@ case class TransferWgsCramV1Metadata(
       // want to fixup files with just '.crai' when possible.
-      craiPath = pathMapper(prefixedCrai)
+      craiPath = movedCram.map { cramUri =>
+        URI.create(s"${cramUri.toString}.crai")
```
~Seems like writing `cramUri + ".crai"` would be simpler here~ actually isn't `toString` unnecessary, e.g. `s"$cramUri.crai"` would work

danxmoran committed:
DSDEGP-1812 Generalize move logic to allow setting new basename.

This is needed for reset-ing filenames after testing a cram delivery
with a patient-ID prefix, and is also useful in general.

danxmoran commented:
Note from #179: if we do end up keeping the hard-coded extensions, it'd be good to expose them as constants so the integration-tests can use them.

tbl3rd commented on clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/TransferMetadata.scala:
```diff
@@ -34,17 +34,17 @@ trait TransferMetadata[M <: TransferMetadata[M]] { self: M =>
     * Until we move to a Path-based API, `destination` must end with '/'.
     */
-  def moveInto(destination: URI, samplePrefix: Option[String] = None): M = {
+  def moveInto(destination: URI, newBasename: Option[String] = None): M = {
```
Probably gotta update the comment here too now.
Maybe for `mapMove()` as well.

tbl3rd reviewed (approved):
Wow. Should have seen this coming. (-Kinda did see this coming, but was swayed by the argument that we should allow only prefixing the sample name!-)

danxmoran committed:
Update comments.

danxmoran committed:
Put expected extensions in constants.

danxmoran commented:
@pshapiro4broad I moved the file extensions into constants.
